pallets/redeem/src/lib.rs

```diff
@@ -917,7 +931,12 @@ impl<T: Config> Pallet<T> {
 
 		ensure!(redeem.vault == vault_id, Error::<T>::UnauthorizedVault);
 
-		let reimbursed_amount = redeem.amount().checked_add(&redeem.transfer_fee())?;
+		//let reimbursed_amount = redeem.amount().checked_add(&redeem.transfer_fee())?;
```

Comment:
Let's remove the comment

Reply:
@TorstenStueber my main concern is that we remove this line. 
@b-yap added changes such that the transferred amount is stored in both cases, 1) when the vault is liquidated and 2) when it's not. The audit ticket only raised concerns about the `transfer_funds_saturated()` function which is used in case 2). 

The existing code calculated the reimbursed_amount with 
```
let reimbursed_amount = redeem.amount().checked_add(&redeem.transfer_fee())?;
```

but in case 1), the amount transferred, (which @b-yap decided to store in the `CancelledRedeemAmount` storage item) is however calculated as
```
let confiscated_collateral = ext::vault_registry::calculate_collateral::<T>(
				&ext::vault_registry::get_liquidated_collateral::<T>(&redeem.vault)?,
				&vault_to_be_burned_tokens,
				&vault_to_be_redeemed_tokens, /* note: this is the value read prior to making
				                               * changes */
			)?;
```

Maybe that does not matter, maybe it does. Anyways, with the changes I suggested in my other comments, we would keep using the existing logic and only store the transferred amount in case 2).

Reply:
The main confusion here arises because the collateral sent from the vault account to the redeemer (`amount_to_slash`) is a collateral amount, however, the `reimbursed_amount` is a wrapped token amount.

The implementation right now is already wrong because `CancelledRedeemAmount` stores the actually transferred collateral amount (the saturated `amount_to_slash`) but uses that as the reimbursement amount in `_mint_tokens_for_reimbursed_redeem` – this is wrong independent of the liquidated/non-liquidated situation we discussed above.

A more fundamental problem already arises in `_cancel_redeem` and not just in `_mint_tokens_for_reimbursed_redeem`: if the redeemer wishes a reimburse and the vault is not undercollateralized (see [this code](https://github.com/pendulum-chain/spacewalk/pull/343/files/f4f92705e0a716f60c924caba5c8e756f5e9a041#diff-066389da98714f72bf9e0f00533ef0829d9181607572e15e866b3285c08841a6R885-R892)), then all the wrapped tokens of the redeemer are immediately transferred to the vault. Also here one could argue that this is unfair to the user if the collateral asset dropped sharply in price when `_cancel_redeem` is executed (i.e., the same argument as in [the audit ticket](https://skyharbor.certik.com/report/1388e1a1-81d0-4c89-812e-c6a3e209bb0c?findingIndex=LI7-04)).

So in order to be completely fair to the user, we should immediately calculate the equivalent wrapped token amount from the actually transferred amount [after the `transfer_funds_saturated`](https://github.com/pendulum-chain/spacewalk/pull/343/files/f4f92705e0a716f60c924caba5c8e756f5e9a041#diff-066389da98714f72bf9e0f00533ef0829d9181607572e15e866b3285c08841a6R849):

```
let transferred_amount = ext::vault_registry::transfer_funds_saturated::<T>(...)
let transferred_amount_in_wrapped =
transferred_amount.saturating_sub(&punishment_fee_in_collateral).convert_to(vault_id.wrapped_currency())?;
```

Then we should:
* for the undercollateralized case: store that value to be used in the `_mint_tokens_for_reimbursed_redeem` function 
* for the non-undercollateralized case: only [send that amount to the vault](https://github.com/pendulum-chain/spacewalk/pull/343/files/f4f92705e0a716f60c924caba5c8e756f5e9a041#diff-066389da98714f72bf9e0f00533ef0829d9181607572e15e866b3285c08841a6R887) account after unlocking

## Liquidated case
This situation is quite different as the [collateral amount sent to the user](https://github.com/pendulum-chain/spacewalk/pull/343/files/f4f92705e0a716f60c924caba5c8e756f5e9a041#diff-066389da98714f72bf9e0f00533ef0829d9181607572e15e866b3285c08841a6R823-R827) (`confiscated_collateral`) is calculated using values that are fixed in the moment the vault turned into the liquidated state. This whole situation assumes that as long as the liquidated state last, the prices don't change. So for that reason I would use the simple solution to set `transferred_amount_in_wrapped` to `vault_to_be_burned_tokens` – exactly as it already is.

Reply:
Are you sure that it should be 
```
let transferred_amount_in_wrapped = transferred_amount.saturating_sub(&punishment_fee_in_collateral).convert_to(vault_id.wrapped_currency())?;
```
and not 
```
let transferred_amount_in_wrapped = transferred_amount.convert_to(vault_id.wrapped_currency())?;
```

Why do we need to subtract the `punishment_fee_in_collateral` from the transferred amount?

Reply:
@TorstenStueber pinging you again on this.

Reply:
I wrote the above answer immediately but saw that it was still pending and not shown to you because I had to submit a review first 🙄.

Reply:
(@ebma)

Reply:
Okay thanks, I think I got it now @TorstenStueber. I think the naming of the variable was throwing me off a bit. 
Maybe we should not call it `transferred_amount_in_wrapped` but `retrievable_amount_in_wrapped`? Or similar, I don't know. But it's not necessarily the transferred amount transformed to wrapped tokens, but less than that because it's what the vault can retrieve later on.

Or maybe it's best to add an extra variable to make it perfectly clear. Something like:
```
let transferred_amount = ext::vault_registry::transfer_funds_saturated::<T>(...)
// Subtract the punishment fee to avoid overpaying the vault when reclaiming tokens
let retrievable_amount_in_collateral = transferred_amount.saturating_sub(&punishment_fee_in_collateral);
let retrievable_amount_in_wrapped = retriveable_amount_in_collateral.convert_to(vault_id.wrapped_currency())?;
```

Reply:
Absolutely, this code is annoyingly complex so I appreciate any effort to make it cleaner and easier to grasp.

Reply:
Perfect. Now that we agreed on the final approach, @b-yap would you be able to implement @TorstenStueber's suggested changes in this PR?

Reply:
Okay so I lost the links after I rebased. I should not have done that prior to reading the comment.

If I understand it right @TorstenStueber, I'll try to rephrase: 
### For the liquidated case:
> So for that reason I would use the simple solution to set transferred_amount_in_wrapped to vault_to_be_burned_tokens – exactly as it already is.

instead of transferring the `confiscated_collateral`, it should be the `vault_to_be_burned_tokens` ?
So we are removing the `confiscated_collateral` in general.
```rust
let vault_to_be_burned_tokens = redeem.amount().checked_add(&redeem.transfer_fee())?;
...
ext::vault_registry::transfer_funds::<T>(
				CurrencySource::LiquidatedCollateral(vault_id.clone()),
				slashing_destination,
				- &confiscated_collateral,
				+ &vault_to_be_burned_tokens  
			)?;
```
^ is this correct? 

When it comes to `fn _mint_tokens_for_reimbursed_redeem()`, it'll be the same amount, correct? 

### For NOT liquidated case:
@ebma Marcel's `retrievable_amount_in_wrapped` is the `amount_to_slash` , AND the amount to use in `fn _mint_tokens_for_reimbursed_redeem()`